src/relay/pass/partial_eval.cc

```diff
@@ -0,0 +1,714 @@
+        }
+        return ConstEvaluate(CallNode::make(expr, args, attrs, type_args), ll);
+      }
+      return ns;
+    };
+  }
+
+  PStatic VisitExpr_(const OpNode* op, LetList* ll) final {
+    return HasStatic(SClos(ConstEvaluateClos(GetRef<Expr>(op), ll)), GetRef<Expr>(op));
+  }
+
+  PStatic VisitExpr_(const ConstructorNode* op, LetList* ll) final {
+    Clos f = [=](const std::vector<PStatic>& pv,
+                 const Attrs& attrs,
+                 const tvm::Array<Type>& type_args,
+                 LetList* ll) {
+      Constructor c = GetRef<Constructor>(op);
+      tvm::Array<Expr> dyn;
+      for (const PStatic& ps : pv) {
+        dyn.push_back(ps->dynamic);
+      }
+      return HasStatic(SConstructor(c, pv), ll->Push(CallNode::make(c, dyn)));
+    };
+    return HasStatic(SClos(f), GetRef<Expr>(op));
+  }
+
+  PStatic VisitExpr_(const MatchNode* op, LetList* ll) final {
+    PStatic ps = VisitExpr(op->data, ll);
+    return env_.Extend<PStatic>([&]() {
+        for (const Clause& c : op->clauses) {
+          switch (VisitPattern(c->lhs, ps)) {
+          case MatchStatus::Match:
+            return VisitExpr(c->rhs, ll);
+          case MatchStatus::NoMatch:
+            continue;
+          case MatchStatus::Unknown:
+            tvm::Array<Clause> clauses;
+            for (const Clause& c : op->clauses) {
+              clauses.push_back(ClauseNode::make(c->lhs,
+                                                 store_.Extend<Expr>([&]() {
+                                                     return VisitExpr(c->rhs, ll)->dynamic;
+                                                   })));
+            }
+            store_.Invalidate();
+            return NoStatic(MatchNode::make(ps->dynamic, clauses));
+          }
+        }
+        LOG(ERROR) << "No case Match";
+        throw;
+      });
+  }
+
+  MatchStatus VisitPattern_(const PatternWildcardNode* op, const PStatic& ps) final {
+    return MatchStatus::Match;
+  }
+
+  MatchStatus VisitPattern_(const PatternVarNode* op, const PStatic& ps) final {
+    env_.Insert(op->var, ps);
+    return MatchStatus::Match;
+  }
+
+  MatchStatus VisitPattern_(const PatternConstructorNode* op, const PStatic& ps) final {
+    if (ps->pstatic) {
+      const SConstructorNode& scn = ps->pstatic->get<SConstructorNode>();
+      CHECK_NE(op->constructor->tag, -1);
+      CHECK_NE(scn.constructor->tag, -1);
+      if (op->constructor->tag == scn.constructor->tag) {
+        // todo(M.K.): should use ptr equality but it is broken
+        CHECK_EQ(op->patterns.size(), scn.fields.size());
+        MatchStatus current_match_status = MatchStatus::Match;
+        for (size_t i = 0; i < op->patterns.size(); ++i) {
+          MatchStatus ms = VisitPattern(op->patterns[i], scn.fields[i]);
+          switch (ms) {
+          case MatchStatus::Match:
+            continue;
+          case MatchStatus::NoMatch:
+            current_match_status = MatchStatus::NoMatch;
+            continue;
```

Comment:
```suggestion
            break;
```
This is an extremely minor optimization that's possible because the match status will never become anything but `NoMatch` after going `NoMatch`